pkg/agent/openflow/pipeline.go

```diff
@@ -1504,125 +1492,13 @@ func (c *client) localProbeFlow(localGatewayIPs []net.IP, category cookie.Catego
-// the SNAT implementation of the external traffic. It is for the Windows Nodes
-// only.
-func (c *client) uplinkSNATFlows(category cookie.Category) []binding.Flow {
-	ctStateNext := dnatTable
-	if c.enableProxy {
-		ctStateNext = endpointDNATTable
-	}
-	bridgeOFPort := uint32(config.BridgeOFPort)
-	flows := []binding.Flow{
-		// Mark the packet to indicate its destination MAC should be rewritten to the real MAC in the L3Forwarding
-		// table, if the packet is a reply to a Pod from an external address.
-		c.pipeline[conntrackStateTable].BuildFlow(priorityHigh).
-			MatchProtocol(binding.ProtocolIP).
-			MatchCTStateNew(false).MatchCTStateTrk(true).
-			MatchCTMark(snatCTMark, nil).
-			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
-			Action().LoadRegRange(int(marksReg), macRewriteMark, macRewriteMarkRange).
-			Action().GotoTable(ctStateNext).
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done(),
-		// Output the non-SNAT packet to the bridge interface directly if it is received from the uplink interface.
-		c.pipeline[conntrackStateTable].BuildFlow(priorityNormal).
-			MatchProtocol(binding.ProtocolIP).
-			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
-			Action().Output(int(bridgeOFPort)).
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done(),
-	}
-	// Forward the IP packets from the uplink interface to
-	// conntrackTable. This is for unSNAT the traffic from the local
-	// Pod subnet to the external network. Non-SNAT packets will be
-	// output to the bridge port in conntrackStateTable.
-	if c.enableProxy {
-		// For the connection which is both applied DNAT and SNAT, the reply packtets
-		// are received from uplink and need to enter CTZoneSNAT first to do unSNAT.
-		//   Pod --> DNAT(CtZone) --> SNAT(CtZoneSNAT) --> ExternalServer
-		//   Pod <-- unDNAT(CtZone) <-- unSNAT(CtZoneSNAT) <-- ExternalServer
-		flows = append(flows, c.pipeline[uplinkTable].BuildFlow(priorityNormal).
-			MatchProtocol(binding.ProtocolIP).
-			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
-			Action().CT(false, conntrackTable, CtZoneSNAT).NAT().CTDone().
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done())
-	} else {
-		flows = append(flows, c.pipeline[uplinkTable].BuildFlow(priorityNormal).
-			MatchProtocol(binding.ProtocolIP).
-			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
-			Action().GotoTable(conntrackTable).
-			Cookie(c.cookieAllocator.Request(category).Raw()).
-			Done())
-	}
-	return flows
-}
-
-// snatFlows installs flows to perform SNAT for traffic to the external network.
-// It is used on Windows Nodes only.
-func (c *client) snatFlows(nodeIP net.IP, localSubnet net.IPNet, category cookie.Category) []binding.Flow {
-	snatIPRange := &binding.IPRange{StartIP: nodeIP, EndIP: nodeIP}
+// snatCommonFlows installs flows to perform SNAT for traffic to the external
```

Comment:
Is it complete? It doesn't seem the flows below perform SNAT

Reply:
snatCommonFlows is for the default common flows on Linux and Windows, which do not really perform SNAT. On Linux only SNAT policies will add flows to look up SNAT IP, mark packets, etc.. Let me revise the comments.